Sources/FoundationEssentials/Calendar/Calendar_Recurrence.swift

```diff
@@ -0,0 +1,835 @@
+                                                          
+                searchLimit = switch recurrence.frequency {
+                    // We want to find a search limit that will stop an infinite
+                    // loop, but that also enumerates through results that users
+                    // might be looking for. Assume that the recurrence requests
+                    // that a date falls on a leap day that occurs on a specific
+                    // weekday, that would yield a result at least once every 40
+                    // years.
+                    case .yearly:   40
+                    case .monthly:  40 * 12
+                    case .weekly:   40 * 53
+                    case .daily:    40 * 366
+                    case .hourly:   40 * 366 * 24
+                    case .minutely: 40 * 366 * 24 * 60
+                }
+            }
+            
+            /// A group of dates within the interval specified by the recurrence
+            /// frequency
+            var currentGroup: [Date] = []
+            
+            /// Calculate the next recurrence without any rules applied
+            mutating func nextBaseRecurrenceDate() -> Date? {
+                func next() -> Date? {
+                    let nextDate: Date?
+                    if iterations == 0 {
+                        // `baseRecurrence` does not include its start date
+                        nextDate = self.start
+                    } else {
+                        nextDate = self.baseRecurrence.next()
+                    }
+                    iterations += 1
+                    return nextDate
+                }
+                while let nextDate = next() {
+                    // Skip every few iterations when an interval has been given
+                    if (iterations - 1) % recurrence.interval != 0 {
+                        continue
+                    }
+                    // If a range has been specified, we should skip a few extra 
+                    // occurrences until we reach the start date
+                    if let rangeLowerBound {
+                        if nextDate < rangeLowerBound {
+                            continue
+                        }
+                    }
+                    return nextDate
+                }
+                return nil
+            }
+            
+            /// Compute `currentGroup` by advancing by a frequency interval from
+            /// the most recent group
+            mutating func nextGroup() {
+                guard let date = self.nextBaseRecurrenceDate() else {
+                    finished = true
+                    return
+                }
+                
+                var dates: [Date] = [date]
+                 
+                // First expand the set of dates, and then filter it
+                for action in [ComponentAction.expand, ComponentAction.limit] {
+                    if monthAction == action {
```

Comment:
Can we add a comment there that the order of expanding or limiting each calendar component needs to stay exactly as this?